circle/src/domain.rs

```diff
@@ -0,0 +1,359 @@
+ .     .
+O       .
+.   .   - (1,0)
+.       O
+ .     .
+  O . .
+
+The full domain is the interleaving of these two cosets
+*/
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct CircleDomain<F> {
+    // log_n corresponds to the size of the WHOLE domain
+    pub(crate) log_n: usize,
+    pub(crate) shift: Complex<F>,
+}
+
+impl<F: ComplexExtendable> CircleDomain<F> {
+    pub(crate) fn standard(log_n: usize) -> Self {
+        Self {
+            log_n,
+            shift: F::circle_two_adic_generator(log_n + 1),
+        }
+    }
+    fn is_standard(&self) -> bool {
+        self.shift == F::circle_two_adic_generator(self.log_n + 1)
+    }
+    pub(crate) fn points(&self) -> impl Iterator<Item = Complex<F>> {
+        let half_gen = F::circle_two_adic_generator(self.log_n - 1);
+        let coset0 = half_gen.shifted_powers(self.shift);
+        let coset1 = half_gen.shifted_powers(half_gen / self.shift);
+        coset0.interleave(coset1).take(1 << self.log_n)
+    }
+
+    #[instrument(skip_all, fields(log_n = %self.log_n))]
+    pub(crate) fn lagrange_basis<EF: ExtensionField<F>>(&self, point: Complex<EF>) -> Vec<EF> {
+        let domain = self.points().collect_vec();
+
+        // the denominator so that the lagrange basis is normalized to 1
+        // this depends only on domain, so should be precomputed
+        let lagrange_normalizer: Vec<F> = domain
+            .iter()
+            .map(|p| s_p_at_p(p.real(), p.imag(), self.log_n))
+            .collect();
+
+        let basis = domain
+            .into_iter()
+            .zip(&lagrange_normalizer)
+            .map(|(p, &ln)| {
+                // ext * base
+                v_0(p.conjugate().rotate(point)) * ln
+            })
+            .collect_vec();
+
+        batch_multiplicative_inverse(&basis)
+    }
+}
+
+impl<F: ComplexExtendable> PolynomialSpace for CircleDomain<F> {
+    type Val = F;
+
+    fn size(&self) -> usize {
+        1 << self.log_n
+    }
+
+    fn first_point(&self) -> Self::Val {
+        point_to_univariate(self.shift).unwrap()
+    }
+
+    fn next_point<Ext: ExtensionField<Self::Val>>(&self, x: Ext) -> Option<Ext> {
+        // Only in standard position do we have an algebraic expression to access the next point.
+        if self.is_standard() {
+            let gen = F::circle_two_adic_generator(self.log_n);
+            Some(point_to_univariate(gen.rotate(univariate_to_point(x))).unwrap())
+        } else {
+            None
+        }
+    }
+
+    fn create_disjoint_domain(&self, min_size: usize) -> Self {
+        let mut log_n = log2_ceil_usize(min_size);
+        // Right now we simply guarantee the domain is disjoint by returning a
+        // larger standard position coset, which is fine because we always ask for a larger
+        // domain. If we wanted good performance for a disjoint domain of the same size,
+        // we could change the shift. Also we could support nonstandard twin cosets.
+        assert!(
```

Comment:
nit: currently, the only public way to build a `CircleDomain` is through the `standard` method. If we don't expose an api for building arbitrary twin domains, we might as well eliminate these checks, as they become redundant.

Reply:
In the future we may add general twin cosets. It could be represented as an enum 

`enum CircleDomain { Standard, General(shift) }`

but in terms of code simplicity this seemed to me like the best way to represent it. There is a very minuscule penalty due to recomputing `circle_two_adic_generator` for each check, but that's fine if the backing implementation is a match statement, otherwise we could have `shift: Option<Complex<F>>`, where `None` is interpreted to mean it's a standard coset, and a getter

`fn get_shift() { self.shift.unwrap_or(F::circle_two_adic_gen(log_n+1)) }`

Reply:
also FYI we might add group position domains at some point, but I haven't thought much about how best to represent that